pageserver/src/pgdatadir_mapping.rs

```diff
@@ -196,6 +198,45 @@ impl<R: Repository> DatadirTimeline<R> {
         Ok(exists)
     }
 
+    pub fn lookup_lsn(
+        &self,
+        timestamp_opt: Option<TimestampTz>,
+        xid_opt: Option<TransactionId>,
+    ) -> Result<Option<(ZTimelineId, Lsn)>> {
+        let lsn = self.tline.get_last_record_lsn();
+        let mut segments: Vec<u32> = self
+            .list_slru_segments(SlruKind::Clog, lsn)?
+            .into_iter()
+            .collect();
+        segments.sort_unstable();
+        for &segno in segments.iter().rev() {
+            let nblocks = self.get_slru_segment_size(SlruKind::Clog, segno, lsn)?;
+            for blknum in (0..nblocks).rev() {
+                let key = slru_block_to_key(SlruKind::Clog, segno, blknum);
+                if let Some((timeline_id, rec_lsn)) =
+                    self.tline.find_record_lsn(key, lsn, &|rec| {
+                        if let ZenithWalRecord::ClogSetCommitted { xids, timestamp } = rec {
+                            if let Some(xact_xid) = xid_opt {
+                                if xids.contains(&xact_xid) {
+                                    return true;
+                                }
+                            }
+                            if let Some(xact_timestamp) = timestamp_opt {
+                                if timestamp <= xact_timestamp {
+                                    return true;
+                                }
+                            }
+                        }
+                        false
+                    })?
+                {
+                    return Ok(Some((timeline_id, rec_lsn)));
```

Comment:
Is this early return correct? Or correct enough? I'd expect xids to be monotonic with key, but maybe timestamps not entirely? If we're accepting some small inacuraccies, a comment would be nice.

Reply:
You are right.
It is not correct: order of XIDs may be different with LSN order.
I will rewrite it.

Reply:
If the solution gets complicated, consider going for the "CLOG index as relish" solution we discussed here https://github.com/zenithdb/zenith/issues/1361. If we downsample the index to second resolution data, it's not that much overhead (there are only 31M seconds in a year). We can continue discussion in the issue if you prefer.

Though if you come up with some quick and easy solution I wouldn't complain :)

Reply:
One more thought: doesn't matter in which order we will traverse CLOG wal recods, in the worst case we have to traverse all PITR interval. So may be optimizing lookup for most recent transactions may be not the right approach.

Just checked: locating LSN of transaction with ~600Mb of original WAL takes about 0.3 seconds.

Concerning maintaining a separate relish: it is not just size overhead, it may cause almost doubling of number of writes and so essential negative effect on performance.